com.unity.sg2/Editor/GraphUI/EditorCommon/Preview/PreviewManager.cs

```diff
@@ -101,7 +99,142 @@ internal void Initialize(ShaderGraphModel graphModel, MainPreviewView mainPrevie
+            m_GraphModel = graphModel;
+            m_MainPreviewView = mainPreviewView;
+
+            ReassignAllNodePreviewTextures();
+        }
+
+        /// <summary>
+        /// After a save or an undo/redo, all the GTF node views get recreated so we gotta go through and reassign any cached node preview textures
+        /// </summary>
+        internal void ReassignAllNodePreviewTextures()
+        {
+            using (var stateUpdater = m_GraphModelStateComponent.UpdateScope)
+            {
+                foreach (var (nodeName, nodeGuid) in m_NodeLookupTable)
+                {
+                    m_GraphModel.TryGetModelFromGuid(nodeGuid, out var nodeModel);
+                    // Don't update the node right away if it is dirty
+                    if (nodeModel is GraphDataNodeModel graphDataNodeModel
+                        && !m_PreviewHandlerInstance.IsNodeDirty(graphDataNodeModel.graphDataName))
+                    {
+                        m_PreviewHandlerInstance.RequestNodePreviewTexture(nodeName, out var nodeRenderOutput, out var shaderMessages, graphDataNodeModel.NodePreviewMode);
+                        graphDataNodeModel.OnPreviewTextureUpdated(nodeRenderOutput);
+
+                        stateUpdater.MarkChanged(graphDataNodeModel);
+                    }
+                }
+            }
+        }
+
+        // Called after an undo/redo to cleanup any stale references to data that no longer exists on graph
+        internal void PostUndoRedoConsistencyCheck()
+        {
+            var keys = m_NodeLookupTable.Keys.ToList();
+            var values = m_NodeLookupTable.Values.ToList();
+            // Remove any nodes from the preview manager that are no longer on graph
+            for(var index = 0; index < m_NodeLookupTable.Count; index++)
+            {
+                var nodeGuid = values[index];
+                var nodeName = keys[index];
+                m_GraphModel.TryGetModelFromGuid(nodeGuid, out var nodeModel);
+                if(nodeModel == null)
+                    OnNodeFlowChanged(nodeName, true);
+            }
+
+            // Add any nodes to the preview manager that are newly on graph
+            foreach (var graphNodeModel in m_GraphModel.NodeModels)
+            {
+                if (!m_NodeLookupTable.ContainsValue(graphNodeModel.Guid))
+                {
+                    if(graphNodeModel is GraphDataNodeModel graphDataNodeModel)
+                        OnNodeAdded(graphDataNodeModel.graphDataName, graphDataNodeModel.Guid);
+
+                    foreach (var edgeModel in graphNodeModel.GetConnectedEdges())
+                    {
+                        if(edgeModel.ToPort.NodeModel is GraphDataNodeModel toPortNodeModel
+                            && toPortNodeModel != graphNodeModel)
+                            OnNodeFlowChanged(toPortNodeModel.graphDataName);
+                    }
+                }
+            }
+        }
```

Comment:
I'm not sure the preview manager should be pushing this stuff down-- but rather have the preview components asking for it when its relevant to their state.